### PR TITLE
Stack vs cabal config files (docs)

### PR DESCRIPTION
**Pull request by snoyberg**
Addresses the common question of why you need to specify dependencies in both stack.yaml and cabal files, giving (hopefully) all of the necessary background.

### CODE REVIEW

File: doc/stack_vs_cabal_config.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+Due to their apparent overlap, the purpose of the following three files can be
+unclear:
+
+* `stack.yaml`
+* A cabal file, e.g. `my-package.cabal`
+* `package.yaml`
+
+The last two are easy to explain: `package.yaml` is a file format supported by
+[hpack](https://github.com/sol/hpack#readme), which adds some niceties on top
+of cabal like YAML syntax support and automatic generation of `exposed-modules`
```
Comment (phadej) [inline doc/stack_vs_cabal_config.md]: More correctly `deb-control` like file: https://www.debian.org/doc/manuals/maint-guide/dreq.en.html

File: doc/stack_vs_cabal_config.md [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+```yaml
+extra-deps:
+- acme-missiles-0.3
+```
+
+What you're saying to Stack is: if at any point you find that you need to build
+the `acme-missiles` package, please use version `0.3`. You are _not_ saying
+"please build `acme-missiles` now." You are also not saying "my package depends
+on `acme-missiles." You are simply making it available should the need arise.
```
Comment (quasicomputational) [inline doc/stack_vs_cabal_config.md]: Missing backtick.

File: doc/stack_vs_cabal_config.md [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+dependency being unavailable.
+
+You may challenge: but why go through all of that annoyance? Stack knows what
+package I want, why not just go grab it? The answer is that, if Stack just
+grabbed `acme-missiles` for you without it being specified in the `stack.yaml`
+somehow, you'd lose reproducibility. How would Stack know which version to use?
+It may elect to use the newest version, but if a new version is available in
+the future, will it automatically switch to that?
+
+Stack's baseline philosoph is that build plans are always reproducible\*. The
```
Comment (quasicomputational) [inline doc/stack_vs_cabal_config.md]: s/philosoph/philosophy/

File: doc/stack_vs_cabal_config.md (hunk 1 above)
Comment (snoyberg) [reply doc/stack_vs_cabal_config.md]: I don't understand this comment, sorry.

File: doc/stack_vs_cabal_config.md (hunk 2 above)
Comment (snoyberg) [reply doc/stack_vs_cabal_config.md]: Good catch, thanks!

File: doc/stack_vs_cabal_config.md (hunk 3 above)
Comment (snoyberg) [reply doc/stack_vs_cabal_config.md]: Another good catch, thanks!

File: doc/stack_vs_cabal_config.md [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+_What's the difference between a `stack.yaml` file and a cabal file?_
+
+## Package versus project
+
+Cabal is a build system, which is used by Stack. Cabal defines the concept of a
+_package_. A package has:
+
+* A name and version
+* 0 or 1 libraries
```
Comment (phadej) [inline doc/stack_vs_cabal_config.md]: There should be at least one library or one executable. Also more correct term would be *exposed libraries*, as `Cabal` does support internal libraries.

File: doc/stack_vs_cabal_config.md [hunk 5]
```diff
@@ -0,0 +1,135 @@
+<div class="hidden-warning"><a href="https://docs.haskellstack.org/"><img src="https://rawgit.com/commercialhaskell/stack/master/doc/img/hidden-warning.svg"></a></div>
+
+# stack vs cabal config files
```
Comment (phadej) [inline doc/stack_vs_cabal_config.md]: The title is confusing. This article doesn't describe `~/.stack/config.yaml` and `~/.cabal/config` differences.

*stack.yaml and pkgname.cabal files*.

To justify *versus*, you could write *stack.yaml vs cabal.project* article, but that's completely different.

File: doc/stack_vs_cabal_config.md (hunk 1 above)
Comment (phadej) [reply doc/stack_vs_cabal_config.md]: Sorry, I missparsed the sentence:

> adds some niceties on top of (cabal like YAML syntax)....

The sentence is too long.

`stack` converts (using [hpack](https://github.com/sol/hpack#readme)) `package.yaml` package descriptions into equivalent `my-package.cabal` files. Hpack is designed to remove redundancies from your package descriptions, for example it uses YAML syntax and supports and automatic generateion of `exposed-modules`

File: doc/stack_vs_cabal_config.md (hunk 5 above)
Comment (snoyberg) [reply doc/stack_vs_cabal_config.md]: I'm having a really difficult time understanding what you're trying to say here. To be honest, I'm failing to understand just about any of the comments you've put on this PR.

File: doc/stack_vs_cabal_config.md (hunk 4 above)
Comment (snoyberg) [reply doc/stack_vs_cabal_config.md]: This is a level of pedantry not needed for understanding this topic.

Comment (phadej) [reply doc/stack_vs_cabal_config.md]: If you aren't pedantic, people *will* misunderstand or misinterpret. I know it's hard to be terse and exact at the same time, but if the goal is to reduce beginner's confusion, we all should pay attention to the docs.

File: doc/stack_vs_cabal_config.md (hunk 5 above)
Comment (phadej) [reply doc/stack_vs_cabal_config.md]: The title of the page is confusing.

- *cabal config* is easily misinterpret as `cabal.config` files. IMHO *cabal package file* is better.
- *stack config* may mean `~/.stack/config.yaml`. IMO *stack.yaml* is better.

TL;DR

```diff
-# stack vs cabal config files
+# stack.yaml vs cabal package files
```

File: doc/stack_vs_cabal_config.md (hunk 4 above)
Comment (snoyberg) [reply doc/stack_vs_cabal_config.md]: OK, but my statement is still _true_, just _incomplete_. I also haven't mentioned that a package needs to specify a build-type or a bunch of other things necessary. What concrete confusion do you expect users to encounter here?
